modAionImpl/src/org/aion/zero/impl/config/dynamic/InFlightConfigReceiver.java

```diff
@@ -0,0 +1,161 @@
+                        key));
+            }
+            Object newVal = getter.apply(newCfg);
+            Object oldVal = getter.apply(activeCfg);
+            if(Objects.equals(newVal, oldVal)) {
+                continue;
+            }
+
+            IDynamicConfigApplier applier = configKeyRegistry.getApplier(key);
+            if(applier == null) {
+                throw new IllegalStateException(String.format(
+                        "DynamicConfigKeyRegistry configuration error.  There is no applier for the bound key '%s'",
+                        key));
+            }
+
+            final InFlightConfigChangeResult result;
+            try {
+                LOG.trace("About to call applier.apply {} for key '{}'", applier, key);
+                result = applier.apply(activeCfg, newCfg);
+            } catch (InFlightConfigChangeException ex) {
+                // if applier.apply worked partially then threw, kernel might be in some broken
+                // state now; might as well try to fix ourselves as much as possible with rollback
+                // though.
+                LOG.error(String.format(
+                        "Error applying change to config key '%s'.  Will attempt rollback.", key), ex);
+                rollback(undoSteps, newCfg);
+                return new ConfigProposalResult(false, ex);
+            }
+
+            LOG.trace("Applier {} returned '{}' for key '{}'", applier, result, key);
+            if(!result.isSuccess()) {
+                LOG.error(String.format(
+                        "Could not apply change to config key '%s'.  Will attempt rollback.", key));
+                rollback(undoSteps, newCfg);
+                return new ConfigProposalResult(false);
+            }
+
+            undoSteps.push(result);
+        }
+        return new ConfigProposalResult(true);
+    }
+
+    private void rollback(Deque<InFlightConfigChangeResult> steps, Cfg newCfg)
+    throws RollbackException {
+        LOG.info("Trying to rollback.  Undo steps are: {}", steps.toString());
+
+        List<InFlightConfigChangeException> exceptions = new LinkedList<>();
+        while(!steps.isEmpty()) {
+            InFlightConfigChangeResult result = steps.pop();
+            try {
+                LOG.trace("About to call undo for application result {}", result);
+                result.getApplier().undo(activeCfg, newCfg);
+            } catch (InFlightConfigChangeException ex) {
+                exceptions.add(ex);
+                LOG.error(String.format("Rollback error while trying to undo %s", result.toString()));
+            }
+        }
+        if(!exceptions.isEmpty()) {
+            throw new RollbackException("Rollback had errors", exceptions);
+        }
+    }
+
+    /**
+     * @return a JMX URL that client can use to connect to JMX server.
+     */
+    public static String createJmxUrl(int port) {
+            return String.format("service:jmx:rmi:///jndi/rmi://127.0.0.1:%d/jmxrmi", port);
+    }
```

Comment:
I have not tried, but it should be possible.  Or at least, the design of JMX is intended to allow you to do that.  

There might be a bit of extra work to make it practical.  For instance, the GUI (JMX client) assumes that the JMX call will be pretty fast, but if it's over a network, it would be significantly slower, so maybe a "Loading" message is needed.

But it's certainly something we can add in the future.

Reply:
Cool, can you write a doc for the current limitations of the GUI?